pygfx/renderers/wgpu/_shadowutil.py

```diff
@@ -21,208 +23,245 @@
     }
 """
 
-binding_layout = [
-    {
-        "binding": 0,
-        "visibility": wgpu.ShaderStage.VERTEX,
-        "buffer": {"type": wgpu.BufferBindingType.uniform},
-    }
-]
+bind_group_layout_entry = {
+    "binding": 0,
+    "visibility": wgpu.ShaderStage.VERTEX,
+    "buffer": {"type": wgpu.BufferBindingType.uniform},
+}
 
 
-class ShadowUtil:
-    def __init__(self, device):
-        self.device = device
-        self.pipelines = {}
-        self.bind_group_layout = self.device.create_bind_group_layout(
-            entries=binding_layout
-        )
+class GlobalShadowState:
+    def __init__(self):
+        self.device = None
+
+    def activate(self, device):
+        if self.device is None:
+            self.device = device
+            self.bind_group_layout = self.device.create_bind_group_layout(
+                entries=[bind_group_layout_entry]
+            )
+
 
-    def get_shadow_pipeline(self, wobject, cull_mode):
-        # shadow pipeline only depends on the object's geometry info (Vertex Buffer)
-        # TODO: now it only depends on the positions, but for instanced meshes, it also needs instance buffer
-        #  Maybe for future skinned meshes , morph buffer is also needed
-        positions = wobject.geometry.positions
+global_shadow_state = GlobalShadowState()
```

Comment:
This global is mainly for the `bind_group_layout`.

We still have a mixed use of `device` - there is a global instance, but we also pass it around in a few placed. The intention for that has been so that in tests we can pass a different instance, but I don't think we actually do this. I'll make a new issue to look into this and refactor things to use `device`  in a uniform way. 

I'll update this code to use the global device.

Reply:
#531

Reply:
mmm, this gets ugly either way with the way things are now. I think this module would be a nice example on how to implement a solution for #531. I'd like to leave it as-is for now.